transport-classes-io_uring/src/main/java/io/netty/channel/uring/AbstractIoUringStreamChannel.java

```diff
@@ -344,20 +391,41 @@ protected void readComplete0(byte op, int res, int flags, short data, int outsta
             final IoUringRecvByteAllocatorHandle allocHandle = recvBufAllocHandle();
             final ChannelPipeline pipeline = pipeline();
             ByteBuf byteBuf = this.readBuffer;
+            IoUringBufferRing bufferRing = lastUsedBufferRing;
             this.readBuffer = null;
-            assert byteBuf != null;
+            this.lastUsedBufferRing = null;
 
             try {
                 if (res < 0) {
                     if (res == Native.ERRNO_ECANCELED_NEGATIVE) {
-                        byteBuf.release();
+                        if (byteBuf != null) {
+                            //recv without buffer ring
+                            byteBuf.release();
+                        }
+                        return;
+                    }
+
+                    if (res == Native.ERRNO_NO_BUFFER_NEGATIVE) {
+                        //recv with provider buffer fail!
+                        //fallback to normal recv
+                        bufferRing.markReadFail();
+                        // fire the BufferRingExhaustedEvent to notify users.
+                        // Users can then switch the ring buffer or do other things as they wish
+                        pipeline.fireUserEventTriggered(new BufferRingExhaustedEvent(bufferRing.bufferGroupId()));
+                        scheduleNextRead(flags);
                         return;
                     }
+
                     // If res is negative we should pass it to ioResult(...) which will either throw
                     // or convert it to 0 if we could not read because the socket was not readable.
                     allocHandle.lastBytesRead(ioResult("io_uring read", res));
                 } else if (res > 0) {
-                    byteBuf.writerIndex(byteBuf.writerIndex() + res);
+                    if (bufferRing != null) {
+                        byteBuf = bufferRing.borrowBuffer(flags >> Native.IORING_CQE_BUFFER_SHIFT, res);
+                        byteBuf.writerIndex(res);
```

Comment:
```suggestion
                        byteBuf.writerIndex(byteBuf.writerIndex() + res);
```

Reply:
I think this byteBuf`s writerIndex is alway 0

Reply:
At the moment yes but it's more of an implementation detail ... so better safe than sorry

Reply:
I agree